[Enhancement]Persist run mode and check after restart

abc982627271 opened the pull request:
- Save run_mode in version file
- Compare the run_mode in version file with the config value on start
- Compare the run_mode in version file with the helper node on first startup

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

kevincai commented on fe/fe-core/src/test/java/com/starrocks/persist/StorageTest.java:
```diff
@@ -113,7 +113,7 @@ public void deleteDir() {
     @Test
     public void testConstruct() {
-        Storage storage1 = new Storage(1, "token", "test");
+        Storage storage1 = new Storage(1, "token", "shared_nothing", "test");
```
add following check?
```java
Assert.assertEquals("shared_nothing", storage1.getRunMode());
```

abc982627271 commented:
run starrocks_fe_unittest

sduzh commented on fe/fe-core/src/main/java/com/starrocks/persist/Storage.java:
```diff
@@ -67,6 +67,7 @@ public class Storage {
     // version file props keys
     private static final String VERSION_PROP_CLUSTER_ID = "clusterId";
     private static final String VERSION_PROP_TOKEN = "token";
+    private static final String VERSION_PROP_RUN_MODE = "run_mode";
```
```suggestion
    private static final String VERSION_PROP_RUN_MODE = "runMode";
```

wyb commented on fe/fe-core/src/main/java/com/starrocks/server/NodeMgr.java:
```diff
@@ -355,6 +371,14 @@ public void getClusterIdAndRoleOnStartup() throws IOException {
+                        runMode = remoteRunMode;
+                        isVersionFileChanged = true;
+                        storage.reload();
+                    }
```
check runMode != null && remoteRunMode != null && runMode != remoteRunMode?

wyb commented on fe/fe-core/src/main/java/com/starrocks/server/NodeMgr.java:
```diff
@@ -272,10 +276,20 @@ public void getClusterIdAndRoleOnStartup() throws IOException {
+
+                if (storage.getRunMode() == null) {
+                    runMode = GlobalStateMgr.getRunMode().name().toLowerCase();
+                    LOG.info("new runMode is {}", runMode);
```
```suggestion
                    LOG.info("new runMode={}", runMode);
```

abc982627271 commented:
run starrocks_fe_unittest

abc982627271 commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 47 (44.68%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|11|35|31.43%|[279, 284, 285, 286, 287, 288, 290, 320, 340, 346, 364, 376, 377, 378, 379, 380, 381, 384, 385, 386, 387, 388, 422, 424]|
|:large_blue_circle:|com/starrocks/persist/Storage.java|9|11|81.82%|[175, 176]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|1|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 21 / 48 (43.75%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|11|35|31.43%|[279, 284, 285, 286, 287, 288, 290, 320, 340, 346, 364, 376, 377, 378, 379, 380, 381, 384, 385, 386, 387, 388, 422, 424]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|[576]|
|:large_blue_circle:|com/starrocks/persist/Storage.java|9|11|81.82%|[175, 176]|

tracymacding commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_fe_unittest

abc982627271 committed:
[Enhancement] Persist run mode and check after restart

- Save run_mode in version
- Compare the run_mode in version file with the config value on start
- Compare the run_mode with helper node

Signed-off-by: abc982627271 <liuxuefen@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 20 / 53 (37.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|13|44|29.55%|[278, 282, 310, 326, 329, 331, 334, 335, 336, 343, 344, 346, 347, 348, 367, 379, 380, 382, 385, 386, 388, 423, 424, 425, 427, 428, 429, 430, 697, 698, 706]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|[576]|
|:large_blue_circle:|com/starrocks/persist/Storage.java|6|7|85.71%|[171]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 20 / 54 (37.04%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/NodeMgr.java|13|45|28.89%|[278, 282, 310, 326, 329, 331, 334, 335, 336, 343, 344, 346, 347, 348, 367, 379, 380, 382, 385, 386, 388, 422, 423, 424, 425, 427, 428, 429, 430, 697, 698, 706]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|[576]|
|:large_blue_circle:|com/starrocks/persist/Storage.java|6|7|85.71%|[171]|

sduzh commented on fe/fe-core/src/main/java/com/starrocks/server/NodeMgr.java:
```diff
@@ -381,13 +413,37 @@ public void getClusterIdAndRoleOnStartup() throws IOException {
+                isVersionFileChanged = true;
+            } else if (RunMode.getCurrentRunMode().isAllowCreateLakeTable()) {
+                LOG.error("Upgrading from a cluster with version less than 3.0 to a cluster with run mode {} of " +
+                          "version 3.0 or above is disallowed. exit exit", RunMode.getCurrentRunMode().getName());
```
```suggestion
                          "version 3.0 or above is disallowed. will exit", RunMode.getCurrentRunMode().getName());
```

sduzh replied:
```suggestion
                          "version 3.0 or above is disallowed. will exit", RunMode.getCurrentRunMode().getName());
```

wanpengfei-git commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_admit_test

sduzh commented:
run starrocks_admit_test

sduzh commented:
run starrocks_admit_test

sduzh commented:
run starrocks_admit_test

tracymacding commented:
run starrocks_admit_test

wanpengfei-git commented:
@Mergifyio backport branch-3.0
